Compose: Remove denominator in character counter

amaranand360 opened the pull request:
Previously, when a message approached or exceeded the character limit, the character counter displayed the format: text_length/max_length. With the changes in this PR, it now shows only the remaining characters and the negative value of exceeded characters, addressing the concerns raised in issue #28706 .

This PR optimizes the `check_overflow_text` function without altering the core logic. It enhances the handling when text.length exceeds max_length by displaying the negative count of  `exceeded characters`. In the else if condition, when text.length reaches 90%, it now shows only the `remaining characters`. The introduction of the `remainingCharacters` variable aids in these cases. This simplification improves clarity without modifying the `check_overflow_text` function logic and showing the remaining characters and negative value of exceeded characters.

- The tooltip associated with the character counter has been updated to display 'Maximum message length: 10000 characters.'
-  Additionally, I removed the banner message, as suggested by the reviewer
- Note: This PR addresses the first two issues in the reported #28706 .

Fixes: #28706

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->



**Screenshots and screen captures:**
- **When Characters exceeded the limit**
<img width="1171" alt="Screenshot 2024-02-15 at 11 44 56 PM" src="https://github.com/zulip/zulip/assets/122713145/862b1dbf-8f32-4367-9314-5f0a1896d79d">





- **Show Remaining Characters**
<img width="1431" alt="Screenshot 2024-02-14 at 10 51 33 PM" src="https://github.com/zulip/zulip/assets/122713145/47a9cda5-fd48-495e-af4a-8e879646f6ce">




https://github.com/zulip/zulip/assets/122713145/39dd68a6-0fd9-4049-bc88-f27a8572e033





<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

You will also need to make sure that all the tests pass.

alya commented:
Thanks! Your commit message still isn't quite right; it would be "compose : Remove denominator in character counter."

Also, it's a critical part of the review process is to clearly communicate the product choices you're making. The banner change you made was not indicated in the issue, and is not clearly documented in your PR description.

alya commented:
Anyway, after playing with the PR:

1. As indicated the issue, we need the explanatory tooltip on the counter itself. Otherwise, it's not even possible to trigger the tooltip until you're over the limit. We can try keeping the tooltip on the "Send" button too, and see how it feels.
1. With this change (and maybe even in general), I think we need to show the counter way later that we currently do. There's no reason to start calling attention to it so early. We could try when there are 900 characters left.
1. I think we can drop the banner altogether. The information is already expressed in the counter + tooltip.

alya commented:
Dropping the banner should be a separate commit.

amaranand360 commented:
@alya I have revised my commit message and updated the PR description to include information about the banner change.

alya commented:
Thanks! Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

N-Shar-ma commented on web/src/compose_validate.js:
```diff
@@ -750,35 +750,26 @@ export function check_overflow_text() {
     // expensive.
     const text = compose_state.message_content();
     const max_length = realm.max_message_length;
+    const remainingCharacters = max_length - text.length;
```
If you observe carefully the other variable names here, you'll see we use snake_case and not camelCase. You can just rename this to `remaining_length` and avoid assigning it values when rendering the template below

amaranand360 replied:
ok I will rename the variable as suggested

Mahhheshh commented on web/src/compose_validate.js:
```diff
@@ -750,35 +750,26 @@ export function check_overflow_text() {
             render_compose_limit_indicator({
-                text_length: text.length,
+                remaining_length: remainingCharacters,
                 max_length,
```
Previously, `max_length` was used in `compose_limit_indicator`. Since it has been removed, there is no need to pass the variable.

amaranand360 replied:
Thank you for pointing that out. I've removed the unused max_length variable.

Mahhheshh commented on web/src/compose_validate.js:
```diff
@@ -750,35 +750,26 @@ export function check_overflow_text() {
             render_compose_limit_indicator({
-                text_length: text.length,
+                remaining_length: remainingCharacters,
                 max_length,
```
same here.

amaranand360 commented:
@N-Shar-ma and @Mahhheshh , I've made the suggested changes, including updating the variable name to remaining_length. and also removed the unused max_length variable.

alya commented:
I thought from the PR description that the feedback has been addressed, but I see in manual testing that these two points from above still need to be handled:

1. As indicated the issue, we need the explanatory tooltip on the counter itself. Otherwise, it's not even possible to trigger the tooltip until you're over the limit. We can try keeping the tooltip on the "Send" button too, and see how it feels.
1. With this change (and maybe even in general), I think we need to show the counter way later that we currently do. There's no reason to start calling attention to it so early. We could try when there are 900 characters left.

N-Shar-ma commented:
@amaranand360 Are you still working on this? I'll take this up if you aren't

amaranand360 commented:
yes, I am actively addressing the points mentioned in review 1. I have resolved those issues, created new commits

N-Shar-ma commented:
@amaranand360 I don't see any updates to the code pushed here since 3 weeks, so I'll take this up now unless you forgot to push changes?

amaranand360 commented:
@N-Shar-ma, I've already implemented the changes based on your review suggestions. Additionally, I've addressed the alternate suggestions. Please review the code, and if there's anything I might have overlooked, kindly let me know. Thank you

N-Shar-ma commented:
> I thought from the PR description that the feedback has been addressed, but I see in manual testing that these two points from above still need to be handled:
> 
>     1. As indicated the issue, we need the explanatory tooltip on the counter itself. Otherwise, it's not even possible to trigger the tooltip until you're over the limit. We can try keeping the tooltip on the "Send" button too, and see how it feels.
> 
>     2. With this change (and maybe even in general), I think we need to show the counter way later that we currently do. There's no reason to start calling attention to it so early. We could try when there are 900 characters left.

@amaranand360 None of Alya's feedback that was posted after your last code push has been addressed

amaranand360 commented:
@N-Shar-ma I've updated the code to address [points](https://github.com/zulip/zulip/pull/28918#issuecomment-1967761692). Please let me know if there's anything else  let me know

N-Shar-ma commented on web/src/compose_validate.ts:
```diff
@@ -700,7 +702,6 @@ export function check_overflow_text(): number {
                 remaining_length,
             }),
         );
-
```
Please avoid making unnecessary changes like this

N-Shar-ma reviewed (changes requested):
The commit message contains unnecessary title case and misses a period at the end, refer to the commit guidelines again

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/29830 for this issue.
